### DISCUSSION TITLE
What's the timeout for a WebQuery?

### DISCUSSION THREAD

**Discussion by jeffpiazza** [arendst/Tasmota#19295] [Q&A]
I monitor my router's wifi connection using approximately the WebQuery form of the rule set given at https://tasmota.github.io/docs/Rules/?query=watchdog#watchdog-for-wi-fi-router-or-modem.

The trigger for the reboot is WebQuery#Data$!Done .  I've looked through the docs, but haven't been able to determine how long after the WebQuery issuance will the "not done" trigger fire.  More importantly, is there a way I can control how long to wait after issuing the WebQuery before concluding that there was no (or an incomplete) response?

**Comment by sfromis** [comment]
With a small test on ESP32, I saw a timeout of 5 seconds before getting `{"WebQuery":"Connect failed"}` when having a server taking longer than that to respond. I've not seen a way to change this, unless you use ESP32 + Berry where webclient has a `set_timeouts` method with a default of 2 seconds.

**Comment by jeffpiazza** [comment]
Does the WebQuery command run to completion (however long that takes) before the next ON clause starts?  If I wanted the router to reboot only if two queries, issued 10 seconds apart, both failed, would this accomplish it?

```
  ON Time#Minute|%var1% DO backlog WebQuery http://192.168.1.10/ GET ENDON
  ON WebQuery#Data$!Done DO backlog Delay 10; WebQuery http://192.168.1.10/ GET ENDON
  ON WebQuery#Data$!Done DO backlog Mult1 3; Power1 0; Delay 10; Power1 1 ENDON
```

**Reply by sfromis** [reply]
No, rules is not some sort of sequence or flow, as `on` is nothing like an `if` from programming languages. Instead, they act as events firing when matched, including triggers for the same event. Additionally, using `Delay` is not reliable for timing, as the interval may be cut short. 

What could work for wanting a retry would be to increment a counter upon failure, and zero it in case of success. Then you could have rules to power off when var2#state==2 and reissue the WebQuery when var2#state==1. Of course, still have the logic for var1 controlling how often to check.
For the delay until powering on again, I can recommend setting `PowerOnState 5` and `PulseTime 100` as this will automatically power on again after each power off.
